torch/_dynamo/variables/misc.py

```diff
@@ -260,6 +293,16 @@ def call_function(self, tx, args, kwargs):
         return AutogradFunctionVariable(self.fn_cls, source=self.source, **options)
 
 
+class SaveSimulatingAutogradFunctionContext(torch.autograd.function.FunctionCtx):
+    def __init__(self):
+        super().__init__()
+        self.saved_tensors = []
+
+    def save_for_backward(self, *tensors: torch.Tensor):
+        super().save_for_backward(tensors)
+        self.saved_tensors.extend([*tensors])
+
```

Comment:
Not sure if this is handled elsewhere, but people are going to call other methods on the ctx object. If they do this, then want to either raise Unsupported (safe default) or model them as well

Reply:
Hmm, well, what we need to model here is what apply cares about, and is not already handled by python. As this is a subclass, we can allow most behavior just fine, I would think? We just need to model anything that gets applied between forward and backward, but not within forward and backward, if that makes sense?

Reply:
Here's a list of context methods that are called reasonably often https://pytorch.org/docs/stable/autograd.html#context-method-mixins, it's fine to not support them and graph break instead, but it's not fine to error out.
These methods can only be applied during forward, not between forward and backward.

Reply:
If they are applied during forward, this will already be handled during tracing time by tracing fwd and installing their guards. During the actual runtime, this is handled by apply.

Reply:
> Answered elsewhere, but maybe we can keep it as is? As long as it stays entire in python in fwd/bwd, it should be fine. save_tensors was special because it gets it's logic done between fwd and bwd, in apply. Alternatively, we could rewrite .apply to do saved_tensors within the context of forward maybe?

Oh I see. Are you saying that when we are speculating the forward, when dynamo sees e.g. ctx.mark_dirty, it will add some guard and then move on?

If that is the case, then yeah, I think we can keep it as is.

Reply:
> this will already be handled during tracing time by tracing fwd and installing their guards. 

So I tried this, and it looks like this doesn't add any guards. But I think this behavior is correct, since the user has no way to change methods on `ctx` (https://gist.github.com/zou3519/1e90d815ad54ca93efbb0a1e97ecbc23), so we don't actually need to add guards that the method is the method we expect. Does that sounds right?

> If that is the case, then yeah, I think we can keep it as is.

Edge case: We probably want to raise Unsupported if the user does ctx.set_materialize_grads(False). Because this will turn any zero grad_output to the backward into a None, and during introspection we explicitly assume that the backwards grad_output are all Tensors.

Everything else, I agree we can keep as-is

Reply:
> during introspection we explicitly assume that the backwards grad_output are all Tensors.

We don't do this for backwards! This is a case we hit in Deberta already, even without set_materialize_grads:
```
    @staticmethod
    def backward(ctx, grad_output):
        if ctx.scale > 1:
            (mask,) = ctx.saved_tensors
            return grad_output.masked_fill(mask, 0) * ctx.scale, None
        else:
            return grad_output, None
```

And we handle this here:
```
            if not isinstance(output, TensorVariable) and not always_restore:
                unimplemented(
                    "HigherOrderOperator with body with non single Tensor output"
                )
```

The idea being, if you soundly got an output, we do not care that it is all tensors or not, since we do not write anything new to the graph here. So this should be safe as well.

Reply:
Oh, I understand. Let me handle this case.

Reply:
Added a test